module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -71,6 +80,346 @@ public function __construct(
+        $dql = 'SELECT IDENTITY(rt.tag) FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'WHERE rt.tag IN (:ids)';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters(compact('ids'));
+        $results = $query->getSingleColumnResult();
+
+        // Record all IDs that are present:
+        $retVal['present'] = array_unique($results);
+
+        // Detect missing IDs:
+        $retVal['missing'] = array_diff($ids, $retVal['present']);
+
+        // Send back the results:
+        return $retVal;
+    }
+
+    /**
+     * Get resources associated with a particular tag.
+     *
+     * @param string $tag  Tag to match
+     * @param string $user ID of user owning favorite list
+     * @param string $list ID of list to retrieve (null for all favorites)
+     *
+     * @return array
+     */
+    public function getResourceIDsForTag($tag, $user, $list = null)
+    {
+        $dql = 'SELECT DISTINCT(rt.resource) AS resource_id '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'WHERE ' . ($this->caseSensitive ? 't.tag = :tag' : 'LOWER(t.tag) = LOWER(:tag) ')
+            . 'AND rt.user = :user ';
+
+        $user = is_object($user) ? $user : $this->entityManager->getReference(User::class, $user);
+        $parameters = compact('tag', 'user');
+        if (null !== $list) {
+            $list = is_object($list) ? $list : $this->entityManager->getReference(UserList::class, $list);
+            $dql .= 'AND rt.list = :list';
+            $parameters['list'] = $list;
+        }
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $result =  $query->getSingleColumnResult();
+        return $result;
+    }
+
+    /**
+     * Get lists associated with a particular tag.
+     *
+     * @param string|array      $tag        Tag to match
+     * @param null|string|array $listId     List ID to retrieve (null for all)
+     * @param bool              $publicOnly Whether to return only public lists
+     * @param bool              $andTags    Use AND operator when filtering by tag.
+     *
+     * @return array
+     */
+    public function getListsForTag(
+        $tag,
+        $listId = null,
+        $publicOnly = true,
+        $andTags = true
+    ) {
+        $tag = (array)$tag;
+        $listId = $listId ? (array)$listId : null;
+        $dql = 'SELECT IDENTITY(rt.list) as list, COUNT(DISTINCT(rt.tag)) AS tagcnt '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'JOIN rt.list l '
+            . 'WHERE rt.resource IS NULL ';
+        $parameters = [];
+        if (null !== $listId) {
+            $dql .= 'AND rt.list IN (:listId) ';
+            $parameters['listId'] = $listId;
+        }
+        if ($publicOnly) {
+            $dql .= 'AND l.public = :public ';
```

Comment:
Since this is a hard-coded value rather than a user-provided variable, would it be better to hard-code the value into the DQL instead of providing it as an additional parameter? I don't mind this approach if you think it best, but all other things being equal, the hard-coded approach would shave off a line of code.